// @atcute/bluesky/dist/limits.js
// this file is generated by scripts/generate-limits.js
// do not edit manually
/** limits for `app.bsky.feed.post` */
export const feedPost = {
    text: { maxGraphemes: 300, maxLength: 3_000 },
    langs: { maxItems: 3 },
    tags: {
        maxItems: 8,
        item: { maxGraphemes: 64, maxLength: 640 },
    },
};
/** limits for `app.bsky.actor.profile` */
export const actorProfile = {
    displayName: { maxGraphemes: 64, maxLength: 640 },
    description: { maxGraphemes: 256, maxLength: 2_560 },
    pronouns: { maxGraphemes: 20, maxLength: 200 },
    avatar: { maxSize: 1_000_000 },
    banner: { maxSize: 1_000_000 },
};
/** limits for `app.bsky.feed.generator` */
export const feedGenerator = {
    displayName: { maxGraphemes: 24, maxLength: 240 },
    description: { maxGraphemes: 300, maxLength: 3_000 },
    avatar: { maxSize: 1_000_000 },
};
/** limits for `app.bsky.feed.threadgate` */
export const feedThreadgate = {
    allow: { maxItems: 5 },
    hiddenReplies: { maxItems: 300 },
};
/** limits for `app.bsky.feed.postgate` */
export const feedPostgate = {
    detachedEmbeddingUris: { maxItems: 50 },
    embeddingRules: { maxItems: 5 },
};
/** limits for `app.bsky.graph.list` */
export const graphList = {
    name: { maxLength: 64, minLength: 1 },
    description: { maxGraphemes: 300, maxLength: 3_000 },
    avatar: { maxSize: 1_000_000 },
};
/** limits for `app.bsky.graph.starterpack` */
export const graphStarterpack = {
    name: { maxGraphemes: 50, maxLength: 500, minLength: 1 },
    description: { maxGraphemes: 300, maxLength: 3_000 },
    feeds: { maxItems: 3 },
};
/** limits for `app.bsky.embed.images` */
export const embedImages = {
    images: { maxItems: 4 },
    image: {
        image: { maxSize: 2_000_000 },
    },
};
/** limits for `app.bsky.embed.video` */
export const embedVideo = {
    video: { maxSize: 100_000_000 },
    captions: { maxItems: 20 },
    alt: { maxGraphemes: 1_000, maxLength: 10_000 },
    caption: {
        file: { maxSize: 20_000 },
    },
};
/** limits for `app.bsky.embed.external` */
export const embedExternal = {
    external: {
        thumb: { maxSize: 1_000_000 },
    },
};
/** limits for `app.bsky.richtext.facet` */
export const richtextFacet = {
    tag: {
        tag: { maxGraphemes: 64, maxLength: 640 },
    },
};
/** limits for `chat.bsky.convo.defs` */
export const convoMessage = {
    text: { maxGraphemes: 1_000, maxLength: 10_000 },
};
