// @atcute/identity/dist/utils.js
import { isHandle } from '@atcute/lexicons/syntax';
import * as t from './types.js';
const isUrlParseSupported = 'parse' in URL;
export const isAtprotoServiceEndpoint = (input) => {
    let url = null;
    if (isUrlParseSupported) {
        url = URL.parse(input);
    }
    else {
        try {
            url = new URL(input);
        }
        catch { }
    }
    return (url !== null &&
        (url.protocol === 'https:' || url.protocol === 'http:') &&
        url.pathname === '/' &&
        url.search === '' &&
        url.hash === '');
};
export const getVerificationMaterial = (doc, id) => {
    const verificationMethods = doc.verificationMethod;
    if (!verificationMethods) {
        return;
    }
    const expectedId = `${doc.id}${id}`;
    for (let idx = 0, len = verificationMethods.length; idx < len; idx++) {
        const { id, type, publicKeyMultibase } = verificationMethods[idx];
        if (id !== expectedId) {
            continue;
        }
        if (publicKeyMultibase === undefined) {
            continue;
        }
        return { type, publicKeyMultibase };
    }
};
export const getAtprotoVerificationMaterial = (doc) => {
    return getVerificationMaterial(doc, '#atproto');
};
export const getAtprotoLabelerVerificationMaterial = (doc) => {
    return getVerificationMaterial(doc, '#atproto_label');
};
export const getAtprotoHandle = (doc) => {
    const alsoKnownAs = doc.alsoKnownAs;
    if (!alsoKnownAs) {
        return null;
    }
    const PREFIX = 'at://';
    for (let idx = 0, len = alsoKnownAs.length; idx < len; idx++) {
        const aka = alsoKnownAs[idx];
        if (!aka.startsWith(PREFIX)) {
            continue;
        }
        const raw = aka.slice(PREFIX.length);
        if (!isHandle(raw)) {
            return undefined;
        }
        return raw;
    }
    return null;
};
export const getAtprotoServiceEndpoint = (doc, predicate) => {
    const services = doc.service;
    if (!services) {
        return;
    }
    for (let idx = 0, len = services.length; idx < len; idx++) {
        const { id, type, serviceEndpoint } = services[idx];
        if (id !== predicate.id && id !== doc.id + predicate.id) {
            continue;
        }
        if (predicate.type !== undefined) {
            if (Array.isArray(type)) {
                if (!type.includes(predicate.type)) {
                    continue;
                }
            }
            else {
                if (type !== predicate.type) {
                    continue;
                }
            }
        }
        if (typeof serviceEndpoint !== 'string' || !isAtprotoServiceEndpoint(serviceEndpoint)) {
            continue;
        }
        return serviceEndpoint;
    }
};
export const getPdsEndpoint = (doc) => {
    return getAtprotoServiceEndpoint(doc, {
        id: '#atproto_pds',
        type: 'AtprotoPersonalDataServer',
    });
};
export const getLabelerEndpoint = (doc) => {
    return getAtprotoServiceEndpoint(doc, {
        id: '#atproto_labeler',
        type: 'AtprotoLabeler',
    });
};
export const getBlueskyChatEndpoint = (doc) => {
    return getAtprotoServiceEndpoint(doc, {
        id: '#bsky_chat',
        type: 'BskyChatService',
    });
};
export const getBlueskyFeedgenEndpoint = (doc) => {
    return getAtprotoServiceEndpoint(doc, {
        id: '#bsky_fg',
        type: 'BskyFeedGenerator',
    });
};
export const getBlueskyNotificationEndpoint = (doc) => {
    return getAtprotoServiceEndpoint(doc, {
        id: '#bsky_notif',
        type: 'BskyNotificationService',
    });
};
