// @atcute/client/dist/client.js
import * as v from '@atcute/lexicons/validations';
import { buildFetchHandler } from './fetch-handler.js';
const JSON_CONTENT_TYPE_RE = /\bapplication\/json\b/;
/** XRPC API client */
export class Client {
    constructor({ handler, proxy = null }) {
        this.handler = buildFetchHandler(handler);
        this.proxy = proxy;
    }
    /**
     * clones this XRPC client
     * @param opts options to merge with
     * @returns the cloned XRPC client
     */
    clone({ handler = this.handler, proxy = this.proxy } = {}) {
        return new Client({ handler, proxy });
    }
    get(name, options = {}) {
        return this.#perform('get', name, options);
    }
    post(name, options = {}) {
        return this.#perform('post', name, options);
    }
    async call(schema, options = {}) {
        // early bailout for tree-shaking when schemas aren't used
        if (!v.xrpcSchemaGenerated) {
            return;
        }
        // extract mainSchema if namespace was passed
        if ('mainSchema' in schema) {
            schema = schema.mainSchema;
        }
        if (schema.params !== null) {
            const paramsResult = v.safeParse(schema.params, options.params);
            if (!paramsResult.ok) {
                throw new ClientValidationError('params', paramsResult);
            }
        }
        if (schema.type === 'xrpc_procedure' && schema.input?.type === 'lex') {
            const inputResult = v.safeParse(schema.input.schema, options.input);
            if (!inputResult.ok) {
                throw new ClientValidationError('input', inputResult);
            }
        }
        const isQuery = schema.type === 'xrpc_query';
        const method = isQuery ? 'get' : 'post';
        if (options.as === undefined && schema.output?.type === 'blob') {
            throw new TypeError(`\`as\` option is required for endpoints returning blobs`);
        }
        const format = options.as !== undefined ? options.as : schema.output?.type === 'lex' ? 'json' : null;
        const response = await this.#perform(method, schema.nsid, {
            params: options.params,
            input: isQuery ? undefined : options.input,
            as: format,
            signal: options.signal,
            headers: options.headers,
        });
        if (format === 'json' && response.ok && schema.output?.type === 'lex') {
            const outputResult = v.safeParse(schema.output.schema, response.data);
            if (!outputResult.ok) {
                throw new ClientValidationError('output', outputResult);
            }
            return {
                ok: true,
                status: response.status,
                headers: response.headers,
                data: outputResult.value,
            };
        }
        return response;
    }
    async #perform(method, name, { signal, as: format = 'json', headers, input, params }) {
        const isWebInput = input &&
            (input instanceof Blob ||
                ArrayBuffer.isView(input) ||
                input instanceof ArrayBuffer ||
                input instanceof ReadableStream);
        const url = `/xrpc/${name}` + _constructSearchParams(params);
        const response = await this.handler(url, {
            method,
            signal,
            body: input && !isWebInput ? JSON.stringify(input) : input,
            headers: _mergeHeaders(headers, {
                'content-type': input && !isWebInput ? 'application/json' : null,
                'atproto-proxy': _constructProxyHeader(this.proxy),
            }),
            duplex: input instanceof ReadableStream ? 'half' : undefined,
        });
        {
            const status = response.status;
            const headers = response.headers;
            const type = headers.get('content-type');
            if (status !== 200) {
                let json;
                if (type != null && JSON_CONTENT_TYPE_RE.test(type)) {
                    // it should be okay to swallow the parsing error here
                    try {
                        const parsed = await response.json();
                        if (isXRPCErrorPayload(parsed)) {
                            json = parsed;
                        }
                    }
                    catch { }
                }
                else {
                    await response.body?.cancel();
                }
                return {
                    ok: false,
                    status: status,
                    headers: headers,
                    data: json ?? {
                        error: `UnknownXRPCError`,
                        message: `Request failed with status code ${status}`,
                    },
                };
            }
            {
                let data;
                switch (format) {
                    case 'json': {
                        if (type != null && JSON_CONTENT_TYPE_RE.test(type)) {
                            // we shouldn't be handling parsing errors
                            data = await response.json();
                        }
                        else {
                            await response.body?.cancel();
                            throw new TypeError(`Invalid response content-type (got ${type})`);
                        }
                        break;
                    }
                    case null: {
                        data = null;
                        await response.body?.cancel();
                        break;
                    }
                    case 'blob': {
                        data = await response.blob();
                        break;
                    }
                    case 'bytes': {
                        data = new Uint8Array(await response.arrayBuffer());
                        break;
                    }
                    case 'stream': {
                        data = response.body;
                        break;
                    }
                }
                return {
                    ok: true,
                    status: status,
                    headers: headers,
                    data: data,
                };
            }
        }
    }
}
// #endregion
// #region Utility functions
const _constructSearchParams = (params) => {
    let searchParams;
    for (const key in params) {
        const value = params[key];
        if (value !== undefined) {
            searchParams ??= new URLSearchParams();
            if (Array.isArray(value)) {
                for (let idx = 0, len = value.length; idx < len; idx++) {
                    const val = value[idx];
                    searchParams.append(key, '' + val);
                }
            }
            else {
                searchParams.set(key, '' + value);
            }
        }
    }
    return searchParams ? `?` + searchParams.toString() : '';
};
const _constructProxyHeader = (proxy) => {
    if (proxy != null) {
        return `${proxy.did}${proxy.serviceId}`;
    }
    return null;
};
const _mergeHeaders = (init, defaults) => {
    let headers;
    for (const name in defaults) {
        const value = defaults[name];
        if (value !== null) {
            headers ??= new Headers(init);
            if (!headers.has(name)) {
                headers.set(name, value);
            }
        }
    }
    return headers ?? init;
};
export const isXRPCErrorPayload = (input) => {
    if (typeof input !== 'object' || input == null) {
        return false;
    }
    const kindType = typeof input.error;
    const messageType = typeof input.message;
    return kindType === 'string' && (messageType === 'undefined' || messageType === 'string');
};
/**
 * takes in the response returned by the client, and either returns the data if
 * it is a successful response, or throws if it's a failed response.
 * @param input either a ClientResponse, or a promise that resolves to a ClientResponse
 * @returns the data from a successful response
 *
 * @example
 * const data = await ok(client.get('com.atproto.server.describeServer'));
 * //    ^? ComAtprotoServerDescribeServer.Output
 */
export const ok = (input) => {
    if (input instanceof Promise) {
        return input.then(ok);
    }
    if (input.ok) {
        return input.data;
    }
    throw new ClientResponseError(input);
};
/** represents an error response returned by the client */
export class ClientResponseError extends Error {
    constructor({ status, headers = new Headers(), data }) {
        super(`${data.error} > ${data.message ?? '(unspecified description)'}`);
        this.name = 'ClientResponseError';
        this.error = data.error;
        this.description = data.message;
        this.status = status;
        this.headers = headers;
    }
}
/** represents a validation error during typed calls */
export class ClientValidationError extends Error {
    constructor(target, result) {
        super(`validation failed for ${target}: ${result.message}`);
        this.name = 'ClientValidationError';
        this.target = target;
        this.result = result;
    }
}
// #endregion
