// @atcute/client/dist/credential-manager.js
import { getPdsEndpoint } from '@atcute/identity';
import { Client, ClientResponseError, isXRPCErrorPayload, ok } from './client.js';
import { simpleFetchHandler } from './fetch-handler.js';
import { decodeJwt } from './utils/jwt.js';
export class CredentialManager {
    /** internal client instance for making authentication requests */
    #server;
    /** holds a promise for the current refresh operation, used for debouncing */
    #refreshSessionPromise;
    /** callback for session expiration */
    #onExpired;
    /** callback for successful session refresh */
    #onRefresh;
    /** callback for session updates */
    #onSessionUpdate;
    constructor({ service, onExpired, onRefresh, onSessionUpdate, fetch: _fetch = fetch, }) {
        this.serviceUrl = service;
        this.fetch = _fetch;
        this.#server = new Client({ handler: simpleFetchHandler({ service, fetch: _fetch }) });
        this.#onRefresh = onRefresh;
        this.#onExpired = onExpired;
        this.#onSessionUpdate = onSessionUpdate;
    }
    /** service URL to make actual API requests with */
    get dispatchUrl() {
        return this.session?.pdsUri ?? this.serviceUrl;
    }
    async handle(pathname, init) {
        await this.#refreshSessionPromise;
        const url = new URL(pathname, this.dispatchUrl);
        const headers = new Headers(init.headers);
        if (!this.session || headers.has('authorization')) {
            return (0, this.fetch)(url, init);
        }
        const initialToken = this.session.accessJwt;
        headers.set('authorization', `Bearer ${initialToken}`);
        const initialResponse = await (0, this.fetch)(url, { ...init, headers });
        if (initialResponse.status !== 401 && !(await isExpiredTokenResponse(initialResponse))) {
            return initialResponse;
        }
        try {
            await this.#refreshSession();
        }
        catch {
            return initialResponse;
        }
        // return initial response if:
        // - request was aborted
        // - refresh failed and cleared the session
        // - token didn't actually change (refresh failed silently)
        // - request body was a stream (can't be resent)
        const updatedToken = this.session?.accessJwt;
        if (init.signal?.aborted ||
            !updatedToken ||
            updatedToken === initialToken ||
            init.body instanceof ReadableStream) {
            return initialResponse;
        }
        // cancel initial response to avoid resource leaks (Node.js)
        await initialResponse.body?.cancel();
        headers.set('authorization', `Bearer ${updatedToken}`);
        return await (0, this.fetch)(url, { ...init, headers });
    }
    #refreshSession() {
        return (this.#refreshSessionPromise ||= this.#refreshSessionInner().finally(() => (this.#refreshSessionPromise = undefined)));
    }
    async #refreshSessionInner() {
        const currentSession = this.session;
        if (!currentSession) {
            return;
        }
        const response = await this.#server.post('com.atproto.server.refreshSession', {
            headers: {
                authorization: `Bearer ${currentSession.refreshJwt}`,
            },
        });
        if (!response.ok) {
            const isExpired = response.status === 401 ||
                response.data.error === 'ExpiredToken' ||
                response.data.error === 'InvalidToken';
            if (isExpired) {
                this.session = undefined;
                this.#onExpired?.(currentSession);
            }
            throw new ClientResponseError(response);
        }
        // DID must not change during refresh
        if (response.data.did !== currentSession.did) {
            this.session = undefined;
            this.#onExpired?.(currentSession);
            throw new ClientResponseError({ status: 401, data: { error: 'InvalidDID' } });
        }
        // protect against concurrent session updates
        if (this.session !== currentSession) {
            throw new Error('concurrent session update detected');
        }
        this.#updateSession({ ...currentSession, ...response.data });
        this.#onRefresh?.(this.session);
    }
    #updateSession(raw) {
        const didDoc = raw.didDoc;
        let pdsUri;
        if (didDoc) {
            pdsUri = getPdsEndpoint(didDoc);
        }
        const newSession = {
            accessJwt: raw.accessJwt,
            refreshJwt: raw.refreshJwt,
            handle: raw.handle,
            did: raw.did,
            pdsUri: pdsUri,
            email: raw.email,
            emailConfirmed: raw.emailConfirmed,
            emailAuthFactor: raw.emailAuthFactor,
            active: raw.active ?? true,
            inactiveStatus: raw.status,
        };
        this.session = newSession;
        this.#onSessionUpdate?.(newSession);
        return newSession;
    }
    /**
     * resume from a persisted session
     * @param session session data, taken from `AtpAuth#session` after login
     */
    async resume(session) {
        // protect against concurrent resume of the same session
        if (session.refreshJwt === this.session?.refreshJwt) {
            await this.#refreshSessionPromise;
            if (!this.session || session.did !== this.session.did) {
                throw new ClientResponseError({ status: 401, data: { error: 'InvalidToken' } });
            }
            return this.session;
        }
        const now = Date.now() / 1_000 + 60 * 5;
        const refreshToken = decodeJwt(session.refreshJwt);
        if (now >= refreshToken.exp || refreshToken.sub !== session.did) {
            throw new ClientResponseError({ status: 401, data: { error: 'InvalidToken' } });
        }
        const accessToken = decodeJwt(session.accessJwt);
        if (accessToken.sub !== session.did) {
            throw new ClientResponseError({ status: 401, data: { error: 'InvalidToken' } });
        }
        // set the session and clear any stale refresh promise
        this.session = session;
        this.#refreshSessionPromise = undefined;
        if (now >= accessToken.exp) {
            // access token expired, need to refresh
            await this.#refreshSession();
        }
        else {
            // access token still valid, fetch session info in background
            const promise = ok(this.#server.get('com.atproto.server.getSession', {
                headers: {
                    authorization: `Bearer ${session.accessJwt}`,
                },
            }));
            promise.then((next) => {
                const existing = this.session;
                if (!existing || existing.did !== next.did) {
                    return;
                }
                this.#updateSession({ ...existing, ...next });
            }, (_err) => {
                // ignore error
            });
        }
        if (!this.session) {
            throw new ClientResponseError({ status: 401, data: { error: 'InvalidToken' } });
        }
        return this.session;
    }
    /**
     * sign in to an account
     * @param options credential options
     * @returns session data
     */
    async login(options) {
        // reset the session
        this.session = undefined;
        this.#refreshSessionPromise = undefined;
        const session = await ok(this.#server.post('com.atproto.server.createSession', {
            input: {
                identifier: options.identifier,
                password: options.password,
                authFactorToken: options.code,
                allowTakendown: options.allowTakendown,
            },
        }));
        return this.#updateSession(session);
    }
    /**
     * sign out of the current session, invalidating it server-side
     */
    async logout() {
        const currentSession = this.session;
        if (!currentSession) {
            return;
        }
        this.session = undefined;
        this.#refreshSessionPromise = undefined;
        try {
            await this.#server.post('com.atproto.server.deleteSession', {
                as: null,
                headers: {
                    authorization: `Bearer ${currentSession.refreshJwt}`,
                },
            });
        }
        catch {
            // ignore errors - session is already cleared locally
        }
    }
}
const isExpiredTokenResponse = async (response) => {
    if (response.status !== 400) {
        return false;
    }
    if (extractContentType(response.headers) !== 'application/json') {
        return false;
    }
    // this is nasty as it relies heavily on what the PDS returns, but avoiding
    // cloning and reading the request as much as possible is better.
    // {"error":"ExpiredToken","message":"Token has expired"}
    // {"error":"ExpiredToken","message":"Token is expired"}
    if (extractContentLength(response.headers) > 54 * 1.5) {
        return false;
    }
    try {
        const data = await response.clone().json();
        if (isXRPCErrorPayload(data)) {
            return data.error === 'ExpiredToken';
        }
    }
    catch { }
    return false;
};
const extractContentType = (headers) => {
    return headers.get('content-type')?.split(';')[0]?.trim();
};
const extractContentLength = (headers) => {
    return Number(headers.get('content-length') ?? ';');
};
